### ISSUE TITLE
States without 2016 Presidential vote

### ISSUE THREAD

**Issue by kuriwaki** [alarm-redist/census-2020#8]
Only two states are missing 2016 Presidential vote: 

- [ ] West Virginia
- [ ] Mississippi. 

This is not really a bug because [VEST](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/NH5S2I), [MGGG](https://github.com/mggg-states), [openelections](https://github.com/openelections/) don't seem to have that data either. But just flagging here in case someone finds a solution. It would be great to have all 50-states.

It is worth noting the NYT Upshot has [2020 precinct results](https://github.com/TheUpshot/presidential-precinct-map-2020) in GeoJSON for those WV and MS. They don't seem to have 2016 for those states, though.

**Comment by rootsmusic** [comment]
Besides those you've linked, another repo for election results is [MEDSL](https://github.com/MEDSL/official-precinct-returns/tree/master/2016-precinct-president).  I haven't explored Mississippi's data, but its 2016 results seem to be at [openelection](https://github.com/openelections/openelections-data-ms/tree/master/2016).

**Comment by rootsmusic** [comment]
WV's 2016 results seem to be at [openelection](https://github.com/openelections/openelections-data-wv/tree/master/2016) too.

**Comment by kuriwaki** [comment]
Thanks. MEDSL and openelections don't have shapefiles; this repo's process matches by geometry. So we will need a precinct shapefile with precinct names that are similar enough to the MEDSL/openelection csv's.

**Comment by rootsmusic** [comment]
[VEST's 2016 results](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/NH5S2I) were mapped on shapefiles.  Hopefully I'm understanding this issue.

**Comment by kuriwaki** [comment]
@rootsmusic: That VEST link does not have a `wv_2016.zip` or a `ms_2016.zip`, which is the root of this issue.

So I'd like to procure those two states from other means.  Like VEST, the product should have a shapefile and 2016 presidential election results on it.  
* MEDSL/open-elections does have election results for those states, for the most part, but not shapefiles. 
* The two states' websites have shapefiles (at least 2010 ones) but no election results. 

To merge the two, we probably need to merge on name of precinct.

**Comment by kuriwaki** [comment]
I looked in a bit more into the data. 

**Mississippi**

* shapefile: ✅  A state agency has a 2010 and 2020 VTD shp file: https://www.maris.state.ms.us/HTML/DATA/Political.html#gsc.tab=0
* MEDSL: ✅ seems to have right number of precincts
* Precinct naming structure: County + name (e.g. county FIPS 28017, `Plesasant Grove`)
* Merge: left_join on  county and precinct links 57% of the 1800 precincts. The rest maybe could be done with careful inspection of the name formats.

**West Virginia**

* shapefile : ❓  I can only find a 2010 VTD shpfile: http://www.wvgis.wvu.edu/data/dataset.php?ID=433 boundaries may or may not have changed since 2010.
* MESDL: ❓ In a few counties they are missing precinct-level results in at least 400 out of ~1800 precincts, and instead report only the county total. Those counties are Kanawha, Cabell (two of the largest counties in the state), Nicholas, Wyoming, Marshall, Webster, Doddridge counties. Fortunately, Openelections WV [data](https://github.com/openelections/openelections-data-wv/tree/master/2016) has csvs of precinct-level election results for all but one those counties. We could stack.
* Precinct naming structure: they use numbers (Kanawha county, 34A), not name
* Merge: Currently, merge on county and precinct number links 66% of the 1800 precincts.

I can try to submit a PR for each state?

**Comment by rootsmusic** [comment]
Precinct shapefiles aren't readily available in many states, and I strongly advise validating even if available.  The Census has [2020 VTD shapefiles](https://www.census.gov/geographies/mapping-files/time-series/geo/tiger-line-file.2020.html) (for P.L. 94-171).  But VTD boundaries should be validated against precinct boundaries, because they may not be equivalent.  If VTD boundaries differ, then hopefully the state (which may not be the board of elections) can provide a crosswalk.

P.S. if you open an issue in MEDSL's repo about their incomplete results, their staff will work on it.

**Comment by kuriwaki** [comment]
Ah yes, Cory and Chris already use the PL94171 as the basis here (via their [PL94171](https://github.com/CoryMcCartan/PL94171) package), and I see that the VTD names in there are similar to what I found each SOS dataset.   So I'll just use the package.

Not sure what "validating" with precinct files would entail though, because I don't have a VEST precinct shapefile. We can certainly check the vote totals add up and we can do a visual inspection with [Upshot's](https://www.nytimes.com/interactive/2018/upshot/election-2016-voting-precinct-maps.html) "Extremely detailed map of 2016".

**Comment by rootsmusic** [comment]
Yes, basic validation is just whether the shapefile is complete.  I advise validating precinct shapefiles because VEST doesn't have time to validate boundaries.  OpenPrecincts isn't active for 2020 elections, but they have a [Python package for verifying shapefiles](https://github.com/OpenPrecincts/verification).

Some of VEST's shapefiles were reconstructed by mixing precincts from different years, under the assumption that boundaries haven't been reshaped.  For states that don't provide a statewide shapefile, VEST's shapefile is constructed by integrating maps from county registrars.  By the way, you may want to ask VEST first whether they're in the process of working on MS and WV.

**Comment by rootsmusic** [comment]
VEST has uploaded both states!

**Comment by christopherkenny** [comment]
Yes, we've already update these along with the elections from KY
